terminus-ui/drawer/src/drawer.component.ts

```diff
@@ -0,0 +1,343 @@
+              public renderer: Renderer2,
+              // tslint:disable-next-line:no-any
+              @Optional() @Inject(DOCUMENT) private doc: any) {
+
+    /**
+     * Listen to `keydown` events outside the zone so that change detection is not run every
+     * time a key is pressed. Re-enter the zone only if the `ESC` key is pressed
+     * and we don't have close disabled.
+     */
+    this.ngZone.runOutsideAngular(() => {
+      (fromEvent(this.elementRef.nativeElement, 'keydown') as Observable<KeyboardEvent>)
+        .pipe(
+          filter(event => event.code === KEYS.ESCAPE.code && !hasModifierKey(event)),
+          untilComponentDestroyed(this)
+        ).subscribe(event => this.ngZone.run(() => {
+          this.collapse();
+          event.stopPropagation();
+          event.preventDefault();
+        }));
+    });
+
+    // We need a Subject with distinctUntilChanged, because the `done` event
+    // fires twice on some browsers.
+    this.animationEnd.pipe(
+      distinctUntilChanged((x, y) => x.fromState === y.fromState && x.toState === y.toState),
+      untilComponentDestroyed(this),
+    ).subscribe((event: AnimationEvent) => {
+      const { fromState, toState } = event;
+
+      if ((toState.indexOf('open') === 0 && fromState === 'void')
+        || (toState === 'void' && fromState.indexOf('open') === 0)
+      ) {
+        this.expandedChange.emit(this.isExpanded);
+      }
+    });
+    this.renderer.setStyle(this.elementRef.nativeElement, 'width', this._expandedSize);
+  }
+
+  public ngAfterContentChecked() {
+    // Enable the animations after the lifecycle hooks have run, in order to avoid animating
+    // drawers that are open by default.
+    if (this.platform.isBrowser) {
+      this.enableAnimations = true;
+    }
+  }
+
+  public ngOnDestroy() {
+    this.modeChanged.complete();
+    this.destroyed.next();
+    this.destroyed.complete();
+  }
+
+  /**
+   * Expand the drawer.
+   *
+   * @return Promise<TsDrawerToggleResult>
+   */
+  public expand(): Promise<TsDrawerToggleResult> {
+    return this.toggle(true);
+  }
+
+  /**
+   * Collapse the drawer.
+   *
+   * @return Promise<TsDrawerToggleResult>
+   */
+  public collapse(): Promise<TsDrawerToggleResult> {
+    this.elementRef.nativeElement.classList.add('ts-drawer--collapsed');
+    return this.toggle(false);
+  }
+
+  /**
+   * Toggle this drawer.
+   *
+   * @param isOpen - whether the drawer should be open.
+   * @return  Promise<TsDrawerToggleResult>
+   */
+  public toggle(isOpen: boolean = !this.isExpanded): Promise<TsDrawerToggleResult> {
+    this._isExpanded = isOpen;
+    if (isOpen) {
+      this.animationState = this.enableAnimations ? 'open' : 'open-instant';
+      // this.elementRef.nativeElement.focus();
+    } else {
+      this.animationState = 'void';
+    }
+
+    return new Promise<TsDrawerToggleResult>(resolve => {
+      this.expandedChange.pipe(take(1)).subscribe(open => resolve(open ? 'open' : 'close'));
+    });
```

Comment:
It doesn't seem like anyone is using this promise.. should the method just return void?

Also, need to add `untilComponentDestroyed`

Reply:
Does it need `untilComponentDestroyed` when it's just `take(1)`?

Reply:
Ah! Nope! I must have blown right by that